Fix inequality(gt & lt) for non lossy types.

sboorlagadda opened the pull request:
For non lossy types inequality is only considered for key lookup direction but
   fails to skip keys that are exact match, so a gt fn is effectively working as ge.
   Fix is to compare the lookup key with provided token and skip if exact match.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/2098)
<!-- Reviewable:end -->

manishrjain commented:
Some comments. Leaving it to @pawanrawal to check this and make the LGTM call.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[worker/tokens.go, line 171 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7EhTTTubOiprNqT6:-L4i7EhTTTubOiprNqT7:b-mq0bg9) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/worker/tokens.go#L171)):*
> ```Go
> 			continue
> 		}
> 		//if its lossy then we handle inequality comparison later on and any custom tokenizer is lossy too
> ```

Over 100 chars.

---

*[worker/tokens.go, line 176 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7FwhfhlmVDD5hXce:-L4i7FwiNt2GlnQmO8Dk:bm9kmoq) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/worker/tokens.go#L176)):*
> ```Go
> 		} else { //for non Lossy lets compare for inequality (gt & lt) to see if key needs to be included
> 			if f == "gt" {
> 				if bytes.Compare([]byte(k.Term), []byte (ineqToken)) > 0 {
> ```

Can you please run gofmt? I think `[]byte<space>(ineqToken)` shouldn't have a space there.

---

*[worker/tokens.go, line 180 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7eUunNBy8I-j7MaE:-L4i7eUunNBy8I-j7MaF:bgbkug0) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/worker/tokens.go#L180)):*
> ```Go
> 				}
> 			} else if f == "lt" {
> 				if bytes.Compare([]byte(k.Term), []byte (ineqToken)) < 0 {
> ```

Can avoid repeated conversion of ineqToken to a byte array, by doing it once outside the loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-:-L4i7jglg7RpeRvl3IoL:b-y56jki)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice fix!

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[query/query_test.go, line 413 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i8Cq95T6IXBoezQ0o:-L4i8Cq95T6IXBoezQ0p:b-wn5dol) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/query/query_test.go#L413)):*
> ```Go
> 	populateGraph(t)
> 	query := `{
> 	  name_starts_with_a(func: lt(name, "B")) {
> ```

name has a bunch of tokenizers. I am not sure if `exact` is chosen for the query. Could you verify if `exact` is used here. If not, then maybe add another predicate and test using that.

---

*[query/query_test.go, line 429 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7KhsNqdyBMQ7Vvq0:-L4i7KhsNqdyBMQ7Vvq1:b-326hnt) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/query/query_test.go#L429)):*
> ```Go
> 	  friends_15_and_17(func: uid(1)) {
> 		name
> 		friend @filter(ge(age, 15) AND lt(age, 18)) {
> ```

Is there anyone with age 18? You might want to modify the test to have a valid value here.

---

*[query/query_test.go, line 436 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7RuFViTvsQx9rOqh:-L4i7RuFViTvsQx9rOqi:b-m4amt3) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/query/query_test.go#L436)):*
> ```Go
> 	}`
> 	js := processToFastJsonNoErr(t, query)
> 	x.Printf("All people: ", js)
> ```

I guess this was only for debugging and can be removed.

---

*[worker/tokens.go, line 171 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7EhTTTubOiprNqT6:-L4i9j2HOwNGURN1c2_b:b-15sc1t) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/worker/tokens.go#L171)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Over 100 chars.
</blockquote></details>

Also please add a space after `//` for all comments.
No need to mention about custom tokenizer. 
You can say later on in `handleCompareAttr` function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-:-L4iBJLIHZnlWCWlafZY:bxpe96o)*
<!-- Sent from Reviewable.io -->

sboorlagadda commented:
Reworked on the comments.

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[query/query_test.go, line 413 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i8Cq95T6IXBoezQ0o:-L4j0nBvXacvbtgg7Ogd:bvhqex6) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/query/query_test.go#L413)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

name has a bunch of tokenizers. I am not sure if `exact` is chosen for the query. Could you verify if `exact` is used here. If not, then maybe add another predicate and test using that.
</blockquote></details>

I did verified that its `exact`, but to avoid any oversight in future I have changed the test to use a different predicate.

---

*[query/query_test.go, line 429 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7KhsNqdyBMQ7Vvq0:-L4j0xsTC5g1vx8Wd6CH:bcmlx1b) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/query/query_test.go#L429)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Is there anyone with age 18? You might want to modify the test to have a valid value here.
</blockquote></details>

Done. There is one with 19, so changed it accordingly.

---

*[query/query_test.go, line 436 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7RuFViTvsQx9rOqh:-L4j10TQheaoHBZ_5vv5:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/query/query_test.go#L436)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

I guess this was only for debugging and can be removed.
</blockquote></details>

Done.

---

*[worker/tokens.go, line 171 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7EhTTTubOiprNqT6:-L4j11XQ_TGQ-tNr_BzI:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/worker/tokens.go#L171)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Also please add a space after `//` for all comments.
No need to mention about custom tokenizer. 
You can say later on in `handleCompareAttr` function.
</blockquote></details>

Done.

---

*[worker/tokens.go, line 176 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7FwhfhlmVDD5hXce:-L4j12PBqoe7j-rV1tQx:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/worker/tokens.go#L176)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Can you please run gofmt? I think `[]byte<space>(ineqToken)` shouldn't have a space there.
</blockquote></details>

Done.

---

*[worker/tokens.go, line 180 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4i7eUunNBy8I-j7MaE:-L4j12v39EHmHObx4D5d:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/7be525973fe5a5525a6414ba2ca15b6ce999a1f0/worker/tokens.go#L180)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Can avoid repeated conversion of ineqToken to a byte array, by doing it once outside the loop.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-:-L4j14FWDoF6dWlliqUB:b-o09s7y)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[query/query_test.go, line 388 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4mSQ2Hx1KVf-nLjSvY:-L4mSQ2Hx1KVf-nLjSvZ:bf7l199) ([raw file](https://github.com/dgraph-io/dgraph/blob/951f3489493e321fee049e9e0e34affb38dcbe69/query/query_test.go#L388)):*
> ```Go
> 
> 	// Data to test inequality (specifically gt, lt) on exact tokenizer
> 	addEdgeToUID(t, "contains", 3000, 3001, nil)
> ```

I don't see these `contains` edges being used.

---

*[worker/tokens.go, line 180 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4mSgKREyFN3CffiyD3:-L4mSgKREyFN3CffiyD4:b-1xqiyu) ([raw file](https://github.com/dgraph-io/dgraph/blob/951f3489493e321fee049e9e0e34affb38dcbe69/worker/tokens.go#L180)):*
> ```Go
> 			// to see if key needs to be included
> 			if f == "gt" {
> 				if bytes.Compare([]byte(k.Term), []byte(ineqTokenInBytes)) > 0 {
> ```

its already in bytes, no need to do `[]byte`

---

*[worker/tokens.go, line 183 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4mSjU_VNzUKB1TBtz4:-L4mSjU_VNzUKB1TBtz5:b-2guby3) ([raw file](https://github.com/dgraph-io/dgraph/blob/951f3489493e321fee049e9e0e34affb38dcbe69/worker/tokens.go#L183)):*
> ```Go
> 					out = append(out, k.Term)
> 				}
> 			} else if f == "lt" {
> ```

same here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098)*
<!-- Sent from Reviewable.io -->

sboorlagadda commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[query/query_test.go, line 388 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4mSQ2Hx1KVf-nLjSvY:-L4mUnOLHkXhilM1xQV7:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/951f3489493e321fee049e9e0e34affb38dcbe69/query/query_test.go#L388)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

I don't see these `contains` edges being used.
</blockquote></details>

Done.

---

*[worker/tokens.go, line 180 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4mSgKREyFN3CffiyD3:-L4mVYjMgPTMcAZsMnV3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/951f3489493e321fee049e9e0e34affb38dcbe69/worker/tokens.go#L180)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

its already in bytes, no need to do `[]byte`
</blockquote></details>

Done.

---

*[worker/tokens.go, line 183 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098#-L4mSjU_VNzUKB1TBtz4:-L4mVWmwbG2m2iXuN66q:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/951f3489493e321fee049e9e0e34affb38dcbe69/worker/tokens.go#L183)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/2098)*
<!-- Sent from Reviewable.io -->
